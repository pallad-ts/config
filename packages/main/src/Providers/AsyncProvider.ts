import { Provider } from "../Provider";

export abstract class AsyncProvider<T> extends Provider<T> {
    private cache?: Promise<Provider.Value<T>>;

    getValue() {
        if (!this.cache) {
            this.cache = this.resolveValueAsync();
        }
        return this.cache;
    }

    abstract resolveValueAsync(): Promise<Provider.Value<T>>;
}
