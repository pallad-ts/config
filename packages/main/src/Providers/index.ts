export * from './DefaultValueProvider';
export * from './EnvProvider';
export * from './FirstAvailableProvider';
export * from './FormattedStringProvider';
export * from './TransformProvider';
export * from './wrapWithDefaultAndTransformer';
