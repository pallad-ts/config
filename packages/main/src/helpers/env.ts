import { envProviderFactory } from "../providersFactory/envProviderFactory";

export const env = envProviderFactory();
