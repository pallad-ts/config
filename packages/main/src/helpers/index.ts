export * from './env';
export * from './format';
export * from './pickByType';
