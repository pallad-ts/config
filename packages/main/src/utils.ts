import {Provider} from './Provider';

export type OptionalPromise<T> = T | Promise<T>;
export type UnwrapProvider<T> = T extends Provider<infer U> ? U : T;
