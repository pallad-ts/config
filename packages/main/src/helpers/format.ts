import {FormattedStringProvider} from '../Providers';

export const format = FormattedStringProvider.create;
