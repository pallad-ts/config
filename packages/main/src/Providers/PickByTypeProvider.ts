import {Provider} from '../Provider';
import {extractProvidersFromConfig} from '../common/extractProvidersFromConfig';
import {Config} from '../Config';
import {runOnOptionalPromise} from '../common/runOnOptionalPromise';
import {ERRORS} from '../errors';
import {isPromise} from '../common/isPromise';
import {loadConfig} from '../common/loadConfig';
import {OptionalPromise} from '../utils';

export class PickByTypeProvider<T> extends Provider<T> {
    private options = new Map<string, any>();

    private constructor(private typeProvider: Provider<string>) {
        super();
    }

    static create(typeDependency: Provider<string>) {
        return new PickByTypeProvider<undefined>(typeDependency);
    }

    registerOptions<TType extends string, TValue>(
        type: TType,
        value: TValue
    ): PickByTypeProvider<NonNullable<T> | PickByTypeProvider.Value<TType, Config.ResolvedValue<TValue>>> {
        if (this.options.has(type)) {
            throw ERRORS.PICK_PROVIDER_TYPE_ALREADY_EXISTS.format(type);
        }
        this.options.set(type, value);
        return this as any;
    }

    private getType(): OptionalPromise<string> {
        return runOnOptionalPromise(
            this.typeProvider.getValue(),
            resolvedType => {
                if (!this.options.has(resolvedType)) {
                    throw ERRORS.PICK_PROVIDER_UNREGISTERED_TYPE.format(
                        resolvedType,
                        Array.from(this.options.keys()).join(', ')
                    );
                }
                return resolvedType;
            }
        );
    }

    private getOptions() {
        return runOnOptionalPromise(this.getType(), type => this.options.get(type)!);
    }

    protected retrieveValue() {
        return runOnOptionalPromise(this.getType(), type => {
            return runOnOptionalPromise(this.getOptions(), options => {
                const finalOptions = Provider.is(options) ? options.getValue() : loadConfig(options);
                return runOnOptionalPromise(finalOptions, result => {
                    return {
                        type,
                        options: result
                    } as any as T;
                })
            });
        });
    }

    getDescription() {
        return `Pick by type from ${this.typeProvider.getDescription()}`;
    }

    private getDependentProviders() {
        return runOnOptionalPromise(
            this.getOptions(),
            options => {
                const optionsDependencies = Provider.is(options) ? [options] : extractProvidersFromConfig(options);
                return [
                    this.typeProvider,
                    ...optionsDependencies
                ];
            }
        );
    }

    isAvailable(): OptionalPromise<boolean> {
        return runOnOptionalPromise<boolean, OptionalPromise<boolean>>(
            this.typeProvider.isAvailable(),
            isTypeAvailable => {
                if (!isTypeAvailable) {
                    return false;
                }

                return runOnOptionalPromise(
                    this.getDependentProviders(),
                    dependentProviders => {
                        const iterator = dependentProviders[Symbol.iterator]();
                        for (const provider of iterator) {
                            const isAvailable = provider.isAvailable();
                            if (isPromise(isAvailable)) {
                                return isAvailable.then(result => {
                                    if (!result) {
                                        return false;
                                    }

                                    return (async () => {
                                        for (const provider of iterator) {
                                            const isAvailable = await provider.isAvailable();
                                            if (!isAvailable) {
                                                return false;
                                            }
                                        }
                                        return true;
                                    })();
                                })
                            } else if (!isAvailable) {
                                return false;
                            }
                        }
                        return true;
                    }
                )
            }
        );
    }
}

export namespace PickByTypeProvider {
    export interface Value<TType extends string, TOptions> {
        type: TType;
        options: TOptions;
    }
}
