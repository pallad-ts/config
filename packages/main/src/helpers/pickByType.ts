import { PickByTypeProvider } from "../Providers/PickByTypeProvider";

export const pickByType = PickByTypeProvider.create;
