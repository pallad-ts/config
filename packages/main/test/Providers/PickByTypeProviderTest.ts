import {env} from '@src/helpers';
import {assert, IsExact} from 'conditional-type-checks';
import {DummyProvider} from '../dummies/DummyProvider';
import {PickByTypeProvider} from '@src/Providers/PickByTypeProvider';
import {OptionalPromise} from '@src/utils';

describe('PickByTypeDependency', () => {

    const TYPE_1 = 'foo';
    const TYPE_2 = 'bar';

    const OPTIONS_1 = {
        foo: 'bar'
    } as const;

    const OPTIONS_2 = {
        bar: 'baz'
    } as const;

    it('description', () => {
        const provider = PickByTypeProvider.create(
            DummyProvider.sync({value: 'foo', description: 'Type description'})
        )
            .registerOptions('foo', DummyProvider.sync({value: OPTIONS_1}))
            .registerOptions('bar', DummyProvider.sync({value: OPTIONS_2}));

        expect(provider.getDescription())
            .toMatchSnapshot();
    });

    describe('registering', () => {
        it('fails if register type that is already registered', () => {
            expect(() => {
                PickByTypeProvider.create(
                    DummyProvider.sync({value: 'test'})
                )
                    .registerOptions('foo', 'foo')
                    .registerOptions('foo', 'bar');
            })
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('sync', () => {
        describe('availability', () => {
            describe('not available if', () => {
                it('type provider is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.syncNotAvailable())
                        .registerOptions('foo', DummyProvider.sync({value: OPTIONS_1}))
                        .registerOptions('bar', DummyProvider.sync({value: OPTIONS_2}));

                    expect(provider.isAvailable())
                        .toBe(false);
                });

                it('options provider is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
                        .registerOptions('foo', DummyProvider.syncNotAvailable())
                        .registerOptions('bar', DummyProvider.syncNotAvailable());

                    expect(provider.isAvailable())
                        .toBe(false);
                });

                it('options has any providers that is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
                        .registerOptions('foo', {
                            foo: DummyProvider.syncNotAvailable()
                        })
                        .registerOptions('bar', {
                            bar: DummyProvider.syncNotAvailable()
                        });

                    expect(provider.isAvailable())
                        .toBe(false);
                });
            });


            it('is available if type and options are available', () => {
                const provider = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
                    .registerOptions('foo', DummyProvider.sync({value: OPTIONS_1}))
                    .registerOptions('bar', DummyProvider.syncNotAvailable());

                expect(provider.isAvailable())
                    .toBe(true);
            });
        });

        describe('getting value', () => {
            it('fails if type is not available', () => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.syncNotAvailable()
                )
                    .registerOptions(TYPE_1, DummyProvider.sync({value: OPTIONS_1}))
                    .registerOptions(TYPE_2, DummyProvider.sync({value: OPTIONS_2}));

                expect(() => {
                    provider.getValue()
                })
                    .toThrowErrorMatchingSnapshot();
            });

            it('fails if resolved type is not registered', () => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.sync({value: 'test'})
                )
                    .registerOptions(TYPE_1, OPTIONS_1)
                    .registerOptions(TYPE_2, OPTIONS_2)

                expect(() => {
                    provider.getValue()
                })
                    .toThrowErrorMatchingSnapshot();
            });

            it.each<[string, any]>([
                [TYPE_1, OPTIONS_1],
                [TYPE_2, OPTIONS_2]
            ])('selects options for given type', (type, expectedOptions) => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.sync({value: type})
                )
                    .registerOptions(TYPE_1, DummyProvider.sync({value: OPTIONS_1}))
                    .registerOptions(TYPE_2, DummyProvider.sync({value: OPTIONS_2}));

                expect(provider.getValue())
                    .toStrictEqual({
                        type,
                        options: expectedOptions
                    });
            });
        });
    });

    describe('async', () => {
        describe('availability', () => {
            describe('not available if', () => {
                it('type provider is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.asyncNotAvailable())
                        .registerOptions('foo', DummyProvider.async({value: OPTIONS_1}))
                        .registerOptions('bar', DummyProvider.async({value: OPTIONS_2}));

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });

                it('options provider is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.async({value: 'foo'}))
                        .registerOptions('foo', DummyProvider.asyncNotAvailable())
                        .registerOptions('bar', DummyProvider.asyncNotAvailable());

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });

                it('options has any providers that is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.async({value: 'foo'}))
                        .registerOptions('foo', {
                            foo: DummyProvider.asyncNotAvailable()
                        })
                        .registerOptions('bar', {
                            bar: DummyProvider.asyncNotAvailable()
                        });

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });
            });


            it('is available if type and options are available', () => {
                const provider = PickByTypeProvider.create(DummyProvider.async({value: 'foo'}))
                    .registerOptions('foo', DummyProvider.async({value: OPTIONS_1}))
                    .registerOptions('bar', DummyProvider.asyncNotAvailable());

                return expect(provider.isAvailable())
                    .resolves
                    .toBe(true);
            });
        });

        describe('getting value', () => {
            it('fails if type is not available', () => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.asyncNotAvailable()
                )
                    .registerOptions(TYPE_1, DummyProvider.async({value: OPTIONS_1}))
                    .registerOptions(TYPE_2, DummyProvider.async({value: OPTIONS_2}));

                return expect(provider.getValue())
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });

            it('fails if resolved type is not registered', () => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.async({value: 'test'})
                )
                    .registerOptions(TYPE_1, OPTIONS_1)
                    .registerOptions(TYPE_2, OPTIONS_2)

                return expect(provider.getValue())
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });

            it.each<[string, any]>([
                [TYPE_1, OPTIONS_1],
                [TYPE_2, OPTIONS_2]
            ])('selects options for given type', (type, expectedOptions) => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.async({value: type})
                )
                    .registerOptions(TYPE_1, DummyProvider.async({value: OPTIONS_1}))
                    .registerOptions(TYPE_2, DummyProvider.async({value: OPTIONS_2}));

                return expect(provider.getValue())
                    .resolves
                    .toStrictEqual({
                        type,
                        options: expectedOptions
                    });
            });
        });
    });

    describe('async + sync', () => {
        describe('availability', () => {
            describe('not available if', () => {
                it('async options provider is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
                        .registerOptions('foo', DummyProvider.asyncNotAvailable())
                        .registerOptions('bar', DummyProvider.asyncNotAvailable());

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });

                it('sync options provider is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.async({value: 'foo'}))
                        .registerOptions('foo', DummyProvider.syncNotAvailable())
                        .registerOptions('bar', DummyProvider.syncNotAvailable());

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });

                it('options has any async providers that is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
                        .registerOptions('foo', {
                            foo: DummyProvider.asyncNotAvailable()
                        })
                        .registerOptions('bar', {
                            bar: DummyProvider.asyncNotAvailable()
                        });

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });

                it('options has any sync providers that is not available', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.async({value: 'foo'}))
                        .registerOptions('foo', {
                            foo: DummyProvider.syncNotAvailable()
                        })
                        .registerOptions('bar', {
                            bar: DummyProvider.syncNotAvailable()
                        });

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(false);
                });
            });

            describe('is available if type and options are available', () => {
                it('async type, sync options', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.async({value: 'foo'}))
                        .registerOptions('foo', DummyProvider.sync({value: OPTIONS_1}))
                        .registerOptions('bar', DummyProvider.syncNotAvailable());

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(true);
                });

                it('sync type, async options', () => {
                    const provider = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
                        .registerOptions('foo', DummyProvider.async({value: OPTIONS_1}))
                        .registerOptions('bar', DummyProvider.syncNotAvailable());

                    return expect(provider.isAvailable())
                        .resolves
                        .toBe(true);
                });
            });
        });

        describe('getting value', () => {
            it('fails if type is not available', () => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.asyncNotAvailable()
                )
                    .registerOptions(TYPE_1, DummyProvider.sync({value: OPTIONS_1}))
                    .registerOptions(TYPE_2, DummyProvider.async({value: OPTIONS_2}));

                return expect(provider.getValue())
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });

            it('fails if resolved type is not registered', () => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.async({value: 'test'})
                )
                    .registerOptions(TYPE_1, OPTIONS_1)
                    .registerOptions(TYPE_2, OPTIONS_2)

                return expect(provider.getValue())
                    .rejects
                    .toThrowErrorMatchingSnapshot();
            });

            it.each<[string, any]>([
                [TYPE_1, OPTIONS_1],
                [TYPE_2, OPTIONS_2]
            ])('selects options for given type', (type, expectedOptions) => {
                const provider = PickByTypeProvider.create(
                    DummyProvider.async({value: type})
                )
                    .registerOptions(TYPE_1, DummyProvider.sync({value: OPTIONS_1}))
                    .registerOptions(TYPE_2, DummyProvider.async({value: OPTIONS_2}));

                return expect(provider.getValue())
                    .resolves
                    .toStrictEqual({
                        type,
                        options: expectedOptions
                    });
            });
        });
    });

    it('types', () => {
        const dep = PickByTypeProvider.create(DummyProvider.sync({value: 'foo'}))
            .registerOptions('foo', DummyProvider.sync<typeof OPTIONS_1>({value: OPTIONS_1}))
            .registerOptions('bar', DummyProvider.sync<typeof OPTIONS_2>({value: OPTIONS_2}))
            .registerOptions('zeta', {test: 'zeta'} as const);

        assert<IsExact<ReturnType<typeof dep.getValue>,
            OptionalPromise<PickByTypeProvider.Value<'foo', typeof OPTIONS_1> |
                PickByTypeProvider.Value<'bar', typeof OPTIONS_2> |
                PickByTypeProvider.Value<'zeta', { test: 'zeta' }>>>>(true);
    });
});
