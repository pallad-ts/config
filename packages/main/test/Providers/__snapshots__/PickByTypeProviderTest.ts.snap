// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PickByTypeDependency async + sync getting value fails if resolved type is not registered 1`] = `"Unregistered type \\"test\\". Registered values: foo, bar"`;

exports[`PickByTypeDependency async + sync getting value fails if type is not available 1`] = `"Config value not available: Pick by type from DummyProvider"`;

exports[`PickByTypeDependency async getting value fails if resolved type is not registered 1`] = `"Unregistered type \\"test\\". Registered values: foo, bar"`;

exports[`PickByTypeDependency async getting value fails if type is not available 1`] = `"Config value not available: Pick by type from DummyProvider"`;

exports[`PickByTypeDependency description 1`] = `"Pick by type from Type description"`;

exports[`PickByTypeDependency registering fails if register type that is already registered 1`] = `"Type \\"foo\\" already registered"`;

exports[`PickByTypeDependency sync getting value fails if resolved type is not registered 1`] = `"Unregistered type \\"test\\". Registered values: foo, bar"`;

exports[`PickByTypeDependency sync getting value fails if type is not available 1`] = `"Config value not available: Pick by type from DummyProvider"`;
